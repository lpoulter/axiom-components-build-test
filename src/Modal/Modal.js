import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Base from '../Base/Base';
import Portal from '../Portal/Portal';
import './Modal.css';

const bodyOpenClassName = 'ax-modal__body--open';

const enableScrolling = () => {
  document.body.classList.remove(bodyOpenClassName);
};

const disableScrolling = () => {
  document.body.classList.add(bodyOpenClassName);
};

export default class Modal extends Component {
  static propTypes = {
    children: PropTypes.node,
    isOpen: PropTypes.bool.isRequired,
    onOverlayClick: PropTypes.func,
    overlayShade: PropTypes.oneOf([
      'shade-1',
      'shade-2',
      'shade-3',
      'shade-4',
    ]),
    overlayTheme: PropTypes.oneOf(['day', 'night']),
    padding: PropTypes.oneOf(['x0', 'x6', 'x8', 'x12', 'x16']),
    shouldCloseOnEsc: PropTypes.bool,
  };

  static defaultProps = {
    overlayShade: 'shade-2',
    padding: 'x6',
  };

  constructor(props) {
    super(props);
    this.handleEscapeKey = this.handleEscapeKey.bind(this);
  }

  componentDidMount() {
    if (this.props.isOpen) {
      disableScrolling();
    }
    if (this.props.shouldCloseOnEsc) {
      document.body.addEventListener('keydown', this.handleEscapeKey, false);
    }
  }

  componentWillUnmount() {
    enableScrolling();
    if (this.props.shouldCloseOnEsc) {
      document.body.removeEventListener('keydown', this.handleEscapeKey, false);
    }
  }

  componentDidUpdate(prevProps) {
    if (!prevProps.isOpen && this.props.isOpen) {
      disableScrolling();
    } else if (prevProps.isOpen && !this.props.isOpen) {
      enableScrolling();
    }

    if (this.props.shouldCloseOnEsc !== prevProps.shouldCloseOnEsc) {
      this.toggleEscapeEventListener();
    }
  }

  toggleEscapeEventListener() {
    if (this.props.shouldCloseOnEsc) {
      document.body.addEventListener('keydown', this.handleEscapeKey, false);
    } else {
      document.body.removeEventListener('keydown', this.handleEscapeKey, false);
    }
  }

  handleEscapeKey(event) {
    const { onOverlayClick, shouldCloseOnEsc } = this.props;

    if (event.key === 'Escape' && shouldCloseOnEsc) {
      if (onOverlayClick) {
        onOverlayClick();
      }
    }
  }

  render() {
    const {
      children,
      isOpen,
      overlayShade,
      overlayTheme,
      onOverlayClick,
      padding,
    } = this.props;

    const classes = classnames('ax-modal__container', `ax-modal__container--padding-${padding}`, {
      [`ax-modal__container--overlay-${overlayShade}`]: overlayShade,
    });

    return isOpen ? (
      <Portal>
        <Base className={ classes } theme={ overlayTheme }>
          { onOverlayClick && (
            <div className="ax-modal__mask" onClick={ onOverlayClick } />
          ) }

          <div className="ax-modal">
            { children }
          </div>
        </Base>
      </Portal>
    ) : null;
  }
}
