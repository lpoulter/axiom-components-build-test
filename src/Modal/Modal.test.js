import React from 'react';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';
import Modal from './Modal';

describe('Modal', () => {
  it('renders', () => {
    const component = mount(<Modal isOpen />);
    expect(toJSON(component)).toMatchSnapshot();
  });

  it('closes if a esc button was pressed', () => {
    const spy = jest.fn();
    const component = mount(<Modal isOpen onOverlayClick={ spy } shouldCloseOnEsc />);
    expect(toJSON(component)).toMatchSnapshot();

    const keyboardEvent = new KeyboardEvent('keydown', { keyCode: 27, key: 'Escape' });
    document.body.dispatchEvent(keyboardEvent);
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
