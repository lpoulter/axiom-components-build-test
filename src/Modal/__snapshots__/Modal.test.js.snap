// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal closes if a esc button was pressed 1`] = `
<Modal
  isOpen={true}
  onOverlayClick={[MockFunction]}
  overlayShade="shade-2"
  padding="x6"
  shouldCloseOnEsc={true}
>
  <Portal>
    <Base
      Component="div"
      className="ax-modal__container ax-modal__container--padding-x6 ax-modal__container--overlay-shade-2"
    >
      <div
        className="ax-modal__container ax-modal__container--padding-x6 ax-modal__container--overlay-shade-2"
      >
        <div
          className="ax-modal__mask"
          onClick={[MockFunction]}
        />
        <div
          className="ax-modal"
        />
      </div>
    </Base>
  </Portal>
</Modal>
`;

exports[`Modal renders 1`] = `
<Modal
  isOpen={true}
  overlayShade="shade-2"
  padding="x6"
>
  <Portal>
    <Base
      Component="div"
      className="ax-modal__container ax-modal__container--padding-x6 ax-modal__container--overlay-shade-2"
    >
      <div
        className="ax-modal__container ax-modal__container--padding-x6 ax-modal__container--overlay-shade-2"
      >
        <div
          className="ax-modal"
        />
      </div>
    </Base>
  </Portal>
</Modal>
`;
